File: tests/arbiter.rs
//! Arbiter service integration tests: a real supervisor on the arbiter
//! fixture house (no broker — the arbiter is a pure bus service), asserting
//! on the forward/preempt/refuse/expiry contract of docs/design.md,
//! Arbitrated mode ("Settled 2026-07-16").

mod common;

use std::time::Duration;

use homeostat::bus::HealthStatus;
use serde_json::{json, Value};
use zenoh::handlers::FifoChannelHandler;
use zenoh::pubsub::Subscriber;

use common::{await_health, health_watch, process_alive, Supervisor};

const FIXTURE: &str = "tests/fixture_house_arbiter";
const CMD_KEY: &str = "home/cmd/hallway/front_door/locked";
const ARBITER_KEY: &str = "home/arbiter/hallway/front_door/locked";
const EVENT_KEY: &str = "home/health/arbiter/event";
const HOLD_MINUTES_KEY: &str = "home/config/arbiter/hold_minutes";

type Sub = Subscriber<FifoChannelHandler<zenoh::sample::Sample>>;

/// Spawns the fixture and waits for the arbiter's liveliness token.
async fn setup() -> (Supervisor, zenoh::Session) {
    let sup = Supervisor::spawn(FIXTURE);
    let observer = sup.observer().await;
    let mut watch = health_watch(&observer, "arbiter").await;
    await_health(&mut watch, Duration::from_secs(60), |h| {
        h.status == HealthStatus::Running
    })
    .await;
    (sup, observer)
}

fn envelope(value: Value, priority: &str, actor: &str) -> Value {
    json!({"value": value, "priority": priority, "actor": actor})
}

async fn put_cmd(session: &zenoh::Session, payload: &Value) {
    session
        .put(CMD_KEY, payload.to_string())
        .await
        .expect("cmd put");
}

/// Writes a parameter through the core's query-with-payload write path
/// (same mechanic as tests/evening.rs's config_write and the dashboard's
/// /api/param).
async fn config_write(session: &zenoh::Session, key: &str, value: Value) -> Value {
    let replies = session
        .get(key)
        .payload(value.to_string())
        .await
        .expect("config write query");
    let reply = replies.recv_async().await.expect("config write reply");
    serde_json::from_slice(&reply.result().expect("write accepted").payload().to_bytes())
        .expect("ok reply is JSON")
}

/// Next sample within the timeout, decoded as JSON.
async fn next_json(sub: &Sub, timeout: Duration) -> Option<Value> {
    let sample = tokio::time::timeout(timeout, sub.recv_async()).await.ok()?.expect("stream open");
    Some(serde_json::from_slice(&sample.payload().to_bytes()).expect("payload is JSON"))
}

/// Asserts no message arrives within the window.
async fn expect_silence(sub: &Sub, window: Duration, what: &str) {
    if let Ok(Ok(sample)) = tokio::time::timeout(window, sub.recv_async()).await {
        panic!(
            "unexpected {what}: {}",
            String::from_utf8_lossy(&sample.payload().to_bytes())
        );
    }
}

/// (a)(b)(c)(d) The full lease lifecycle on one arbitrated entity: an
/// automation wish forwards and takes the lease; a manual wish preempts it
/// (event fired) and forwards; a subsequent automation wish is refused (no
/// forward, event fired); shrinking hold_minutes to a sub-second value and
/// waiting it out reopens the entity, so automation wishes flow again.
#[tokio::test(flavor = "multi_thread")]
async fn forward_preempt_refuse_and_expiry() {
    let (mut sup, observer) = setup().await;
    let arbiter_sub = observer
        .declare_subscriber(ARBITER_KEY)
        .await
        .expect("arbiter subscriber");
    let event_sub = observer
        .declare_subscriber(EVENT_KEY)
        .await
        .expect("event subscriber");

    // (a) No holder yet: an automation wish forwards unchanged and takes
    // the lease at the automation band.
    let auto_wish = envelope(json!(true), "automation", "scheduler");
    put_cmd(&observer, &auto_wish).await;
    let forwarded = next_json(&arbiter_sub, Duration::from_secs(10))
        .await
        .expect("automation wish forwarded");
    assert_eq!(forwarded, auto_wish, "envelope forwarded unchanged");
    expect_silence(&event_sub, Duration::from_millis(500), "event on a clean take").await;

    // Shrink hold_minutes now, before the manual takeover below: the next
    // lease taken (by the manual wish) is computed from the new value, so
    // it will expire in well under a second once we wait it out in (d).
    let written = config_write(&observer, HOLD_MINUTES_KEY, json!(0.01)).await;
    assert_eq!(written, json!(0.01));

    // (b) A manual wish preempts the still-active (strictly lower)
    // automation holder: preempt event, then the forward.
    let manual_wish = envelope(json!(false), "manual", "owner");
    put_cmd(&observer, &manual_wish).await;
    let event = next_json(&event_sub, Duration::from_secs(10))
        .await
        .expect("preempt event");
    assert_eq!(
        event,
        json!({
            "kind": "preempt",
            "room": "hallway",
            "entity": "front_door",
            "aspect": "locked",
            "from_priority": "automation",
            "from_actor": "scheduler",
            "to_priority": "manual",
            "to_actor": "owner",
        })
    );
    let forwarded = next_json(&arbiter_sub, Duration::from_secs(10))
        .await
        .expect("manual wish forwarded");
    assert_eq!(forwarded, manual_wish, "envelope forwarded unchanged");

    // (c) An automation wish right afterwards is refused: the manual lease
    // (just taken, hold_minutes now 0.01 = 0.6s) is still active and
    // strictly higher. No forward; a refuse event instead.
    put_cmd(&observer, &auto_wish).await;
    let event = next_json(&event_sub, Duration::from_secs(10))
        .await
        .expect("refuse event");
    assert_eq!(
        event,
        json!({
            "kind": "refuse",
            "room": "hallway",
            "entity": "front_door",
            "aspect": "locked",
            "priority": "automation",
            "actor": "scheduler",
            "holder_priority": "manual",
            "holder_actor": "owner",
        })
    );
    expect_silence(&arbiter_sub, Duration::from_millis(500), "forward of a refused wish").await;

    // (d) Once the shrunk lease's deadline passes, the entity reopens: the
    // same automation wish now forwards again, taking a fresh lease.
    tokio::time::sleep(Duration::from_millis(900)).await;
    put_cmd(&observer, &auto_wish).await;
    let forwarded = next_json(&arbiter_sub, Duration::from_secs(10))
        .await
        .expect("automation wish forwards again after expiry");
    assert_eq!(forwarded, auto_wish);
    expect_silence(&event_sub, Duration::from_millis(500), "event on an expiry re-take").await;

    sup.shutdown();
}

/// (f) A malformed cmd envelope for an arbitrated entity drops with an
/// "invalid-command" health event, and never reaches home/arbiter/**.
#[tokio::test(flavor = "multi_thread")]
async fn malformed_envelope_drops_with_health_event() {
    let (mut sup, observer) = setup().await;
    let arbiter_sub = observer
        .declare_subscriber(ARBITER_KEY)
        .await
        .expect("arbiter subscriber");
    let event_sub = observer
        .declare_subscriber(EVENT_KEY)
        .await
        .expect("event subscriber");

    // A bare value (the pre-envelope shape) is not an envelope.
    observer.put(CMD_KEY, "true").await.expect("cmd put");
    let event = next_json(&event_sub, Duration::from_secs(10))
        .await
        .expect("invalid-command event");
    assert_eq!(event["kind"], json!("drop"));
    assert_eq!(event["reason"], json!("invalid-command"));
    expect_silence(&arbiter_sub, Duration::from_millis(1500), "forward of a malformed command")
        .await;

    // An envelope with an unknown priority is just as malformed.
    observer
        .put(CMD_KEY, json!({"value": true, "priority": "urgent", "actor": "x"}).to_string())
        .await
        .expect("cmd put");
    let event = next_json(&event_sub, Duration::from_secs(10))
        .await
        .expect("invalid-command event");
    assert_eq!(event["reason"], json!("invalid-command"));

    // Still alive and translating afterwards.
    let ok_wish = envelope(json!(true), "automation", "scheduler");
    put_cmd(&observer, &ok_wish).await;
    let forwarded = next_json(&arbiter_sub, Duration::from_secs(10))
        .await
        .expect("a well-formed wish still forwards");
    assert_eq!(forwarded, ok_wish);

    sup.shutdown();
}

/// (e) The arbiter honors the step-2 unit contract: liveliness token when
/// ready, clean SIGTERM shutdown within the grace, no orphans.
#[tokio::test(flavor = "multi_thread")]
async fn adapter_honors_unit_contract() {
    let (mut sup, observer) = setup().await;
    let mut watch = health_watch(&observer, "arbiter").await;
    let health = await_health(&mut watch, Duration::from_secs(10), |h| {
        h.status == HealthStatus::Running
    })
    .await;
    let pid = health.pid.expect("running unit has a pid");
    assert!(process_alive(pid), "arbiter alive before shutdown");

    sup.signal(libc::SIGTERM);
    // shutdown_grace_s = 5 in the fixture; a graceful exit must fit inside
    // it with margin only for reaping and bus teardown.
    let code = sup.wait_exit(Duration::from_secs(7));
    assert_eq!(code, Some(0), "supervisor exit code");
    assert!(!process_alive(pid), "arbiter must not outlive the supervisor");
}

File: adapters/arbiter.py
# /// script
# requires-python = ">=3.11"
# dependencies = [
#     "homeostat",
# ]
#
# [tool.uv.sources]
# homeostat = { path = "../sdk/python", editable = true }
# ///
"""Arbiter service: the write-token holder for arbitrated entities (see
docs/design.md, Arbitrated mode, "Settled 2026-07-16").

By plan-time construction an adapter's templated cmd subscription excludes
its arbitrated entities, so wishes for them never reach an owner adapter
directly; this service subscribes every wish on home/cmd/** instead. A
wish for a non-arbitrated entity is ignored — its own adapter consumes
home/cmd directly. A wish for an arbitrated entity holds a lease per
(entity, aspect) — the granularity of the cmd key itself, amended from
per-entity when the heat pump showed orthogonal control dimensions
sharing one entity (the family's setpoint must not freeze an
automation's outdoor_temperature_offset; see docs/design.md, Arbitrated
mode): {priority, actor, deadline}, deadline
`time.monotonic() + hold_minutes * 60`. No active lease, an expired one,
or an incoming priority at or above the holder's band (band order
keys.CMD_PRIORITIES, manual highest — THE FAMILY ALWAYS WINS OVER
AUTOMATIONS) forwards the envelope unchanged to
home/arbiter/{room}/{entity}/{aspect} (keys.arbiter_key) and takes or
refreshes the lease at the incoming band/actor; a takeover from a
strictly lower active holder additionally publishes a "preempt" event; an
incoming priority strictly below the holder is refused with a "refuse"
event and no forward. Expiry reopens the aspect to automations, so a
forgotten override self-heals. A malformed envelope drops with an
"invalid-command" event, like any adapter. Events land at
home/health/arbiter/event, recorded like any health event.

What it is holding is published as one document per arbiter unit at
home/hold/{unit}, the home/discovery/{unit} shape, mirrored by the core.
A lease is the answer to "is this aspect held right now?", which the
preempt/refuse events cannot give: they are an audit trail, and a consumer
that joins mid-hold has missed them. One document rather than a key per
lease, because that is what makes the awkward cases trivial — a restart
publishes an empty list (the leases were memory and are gone), expiry needs
one timer rather than one per lease, and a reader sees a consistent set.

Enforcement stays on time.monotonic(), which no clock step can shorten or
stretch; the published `until` is its wall-clock twin, because a countdown
is the only kind of deadline that means anything in another process. The
two can disagree if the clock is stepped, and only the countdown suffers.

hold_minutes is a family-editable parameter, kept live by the SDK's
LiveParams (subscribe-then-get, seeded from the manifest default) rather
than routed through automation.Context, whose Context.publish only knows
the state/cmd key-slot shape and has no notion of this service's
arbitrary, per-wish home/arbiter/{room}/{entity}/{aspect} forwarding keys.
"""

import datetime
import json
import os
import signal
import threading
import time

import homeostat
from homeostat import house, keys
from homeostat.params import LiveParams

PARAM = "hold_minutes"


class Params(LiveParams):
    """hold_minutes from home/config/{unit}/*, live."""

    @property
    def hold_minutes(self) -> float:
        return self.get(PARAM)


def iso(epoch: float) -> str:
    """An epoch second as RFC3339 UTC, the spelling every other timestamp
    on this bus uses."""
    return (
        datetime.datetime.fromtimestamp(epoch, datetime.timezone.utc)
        .isoformat(timespec="seconds")
        .replace("+00:00", "Z")
    )


def main():
    unit = os.environ[keys.ENV_UNIT]
    model = house.load_house(".")
    arbitrated = {(e.room, e.name) for e in model.entities if e.write_mode == "arbitrated"}
    own = next(u for u in model.units if u.name == unit)

    session = homeostat.connect()
    params = Params(session, {PARAM: float(own.params[PARAM]["default"])})
    lock = threading.Lock()
    leases: dict[tuple[str, str, str], dict] = {}
    hold_key = keys.hold_key(unit)
    # Wakes the expiry thread when the earliest deadline moves — a new
    # lease, or one pruned — so it never sleeps past a hold's end.
    changed = threading.Condition(lock)

    def publish_holds_locked() -> None:
        """The leases still in force, oldest first. Called with the lock
        held, from whoever changed them; the document replaces its
        predecessor, so a consumer never merges two of them."""
        now = time.monotonic()
        wall = time.time()
        holds = []
        for (room, entity, aspect), lease in sorted(leases.items()):
            if lease["deadline"] <= now:
                continue
            holds.append(
                {
                    "room": room,
                    "entity": entity,
                    "aspect": aspect,
                    "priority": lease["priority"],
                    "actor": lease["actor"],
                    "since": iso(wall - (now - lease["taken"])),
                    # The monotonic deadline is what arbitration enforces;
                    # this is what a countdown can be drawn from.
                    "until": iso(wall + (lease["deadline"] - now)),
                    # How many wishes this hold has actually refused: not
                    # what makes it a hold, but what says it cost something.
                    "refused": lease["refused"],
                }
            )
        session.put_json(hold_key, {"schema": 1, "holds": holds})

    def prune_locked() -> bool:
        """Drops expired leases. True when something went, so the caller
        knows whether the document changed."""
        now = time.monotonic()
        expired = [k for k, lease in leases.items() if lease["deadline"] <= now]
        for k in expired:
            del leases[k]
        return bool(expired)

    def expiry_loop(stop: threading.Event) -> None:
        """One thread for every lease, waiting on the earliest deadline.
        Expiry is otherwise lazy — evaluated on the next wish for that key —
        which is correct for arbitration and wrong for a published
        document, which would go on claiming a hold that had ended."""
        while not stop.is_set():
            with changed:
                if prune_locked():
                    publish_holds_locked()
                deadlines = [lease["deadline"] for lease in leases.values()]
                now = time.monotonic()
                timeout = max(0.05, min(deadlines) - now) if deadlines else None
                changed.wait(timeout=timeout)

    def cmd_handler(sample):
        key = str(sample.key_expr)
        parts = key.split("/", 4)
        if len(parts) < 5:
            session.health_event("drop", reason="off-schema-key", key=key)
            return
        room, entity, aspect = parts[2:5]
        if (room, entity) not in arbitrated:
            return  # not arbitrated: its own adapter consumes this wish

        envelope = None
        try:
            envelope = json.loads(sample.payload.to_bytes())
            keys.parse_cmd_envelope(envelope)
            priority, actor = envelope["priority"], envelope["actor"]
            if not isinstance(actor, str):
                raise ValueError("cmd envelope actor is not a string")
        except (ValueError, KeyError):
            # A payload that never parsed has no id; one that parsed but is
            # not an envelope may still carry the id its publisher is
            # waiting on, and cmd_envelope_id takes either.
            session.health_event(
                "drop", reason="invalid-command", key=key, cmd_id=keys.cmd_envelope_id(envelope)
            )
            return
        cmd_id = keys.cmd_envelope_id(envelope)

        incoming = keys.CMD_PRIORITIES.index(priority)
        with changed:
            now = time.monotonic()
            lease = leases.get((room, entity, aspect))
            holder = lease if lease is not None and now < lease["deadline"] else None
            if holder is not None and incoming < keys.CMD_PRIORITIES.index(holder["priority"]):
                action = "refuse"
                # What the hold cost, carried on the hold itself: a
                # consumer can say "this override has turned an automation
                # away twice" without replaying the event log.
                holder["refused"] += 1
            else:
                action = (
                    "preempt"
                    if holder is not None
                    and incoming > keys.CMD_PRIORITIES.index(holder["priority"])
                    else "forward"
                )
                leases[(room, entity, aspect)] = {
                    "priority": priority,
                    "actor": actor,
                    "deadline": now + params.hold_minutes * 60,
                    "taken": now,
                    # A refreshed hold starts its tally again: the count
                    # belongs to the hold in force, not to the aspect.
                    "refused": 0,
                }
            publish_holds_locked()
            # The earliest deadline has moved either way — a fresh lease
            # pushes it out, and the expiry thread must not sleep on the
            # old one.
            changed.notify_all()

        if action == "refuse":
            # The one outcome that is neither success nor failure: the
            # command was well-formed and reached the arbiter, and a higher
            # band simply holds the aspect. `cmd_id` is what lets whoever
            # published it say so, instead of waiting out a timeout it was
            # never going to win.
            session.health_event(
                "refuse",
                room=room,
                entity=entity,
                aspect=aspect,
                priority=priority,
                actor=actor,
                cmd_id=cmd_id,
                holder_priority=holder["priority"],
                holder_actor=holder["actor"],
            )
            return
        if action == "preempt":
            session.health_event(
                "preempt",
                room=room,
                entity=entity,
                aspect=aspect,
                from_priority=holder["priority"],
                from_actor=holder["actor"],
                to_priority=priority,
                to_actor=actor,
            )
        session.put_json(keys.arbiter_key(room, entity, aspect), envelope)

    cmd_sub = session.subscribe("home/cmd/**", cmd_handler)

    # Leases are memory, so a restart holds nothing — and the mirror still
    # carries whatever the last process published. One empty document says
    # so exactly, which is the whole reason this is one key and not one per
    # hold: there is no set to enumerate and clear.
    with changed:
        publish_holds_locked()

    session.ready()

    stop = threading.Event()
    signal.signal(signal.SIGTERM, lambda *_: stop.set())
    signal.signal(signal.SIGINT, lambda *_: stop.set())
    expiry = threading.Thread(target=expiry_loop, args=(stop,), daemon=True)
    expiry.start()
    stop.wait()

    with changed:
        changed.notify_all()  # let the expiry thread see the stop flag
    expiry.join(timeout=2)
    cmd_sub.undeclare()
    session.close()


if __name__ == "__main__":
    main()
